### PR TITLE
Fix `possibly-used-before-assignment` issues in pylint codebase

### LINKED ISSUE #9637: Fix `possibly-used-before-assignment` issues in our own codebase
This new check was turned off for pylint itself with the intent to just get the check out the door. A PR to improve the pylint codebase and enable the check would be welcome.

### PR DESCRIPTION
**Pull request by aatle**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

This new check was disabled for pylint itself. Enabled, it gives 7 errors in 5 files in the codebase.

Enables the check for pylint and fixes the 7 errors.

Three of the errors were fixed by adding an `else` clause that explicitly raised `ValueError` for bad values instead of letting the code possibly run into an `UnboundLocalError`.
Two errors were caused by checking a bool flag (that never changes) in two separate conditionals: one where the variable set, the other where it was used.
The last two errors were caused by relying on a separate flag when setting a variable, for the conditional that uses the variable.

Closes #9637

<sub> Note: this is my first contribution to open source, so feedback and advice are welcome </sub>

### CODE REVIEW

**Comment by Pierre-Sassoulas** [comment]
Hey @aatle thank you for your contribution ! It would be better if each commit was passing the tests (i.e. first commit activates the check and disable warnings locally, then each following commit remove a local disable and fix the associated warning). That way we can look at each individual commit during review and partially revert or drop them later. You can open a merge request for each too if necessary. (Sometime it's nice to move what's consensual out of the way.)

File: pylint/checkers/__init__.py [hunk 1]
```diff
@@ -109,6 +109,8 @@ def table_lines_from_stats(
                 ("warning", "NC"),
                 ("error", "NC"),
             ]
+    else:
+        raise ValueError(stat_type)
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/__init__.py:113]: I think I'd rather have default value like 
```python
      [
                ("convention", "NC"),
                ("refactor", "NC"),
                ("warning", "NC"),
                ("error", "NC"),
            ]
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
Could you please open a first merge request with https://github.com/pylint-dev/pylint/pull/9644/commits/bea0aae17bec65468aceccdb8af94702b3732f51 and the new local disable first and the commit with no value error raised like 9e8d43f8bf23861fcfbfa53ba5343239e7ddb542 ? I think those change are mergeable as is This will make you trusted enough to run the CI without our approval and ease both of our lives and also make the following review smaller :)

**Comment by DanielNoord** [comment]
@aatle does this need a rebase?

**Comment by aatle** [comment]
@DanielNoord I am not really sure (I'm not that experienced).
I had created and merged a new pull request which includes most of the content of this pull request because of the code coverage test failing for this one. The other uses ignores instead of raising ValueError in 3 cases.

I would think the best thing to do is to close this pull request and possibly the issue too. Merging this pull request wouldn't be that great, a new pull request to make these changes (if desired) would probably be better.

If you agree, this PR can be closed and probably the original issue too, as the warnings have all been fixed (mostly with ignores).

**Comment by DanielNoord** [comment]
I'm fine with closing this, but I'm interested to see what is left after rebasing on `main`. I think Pierre already gave some pointers to remove some of the `ValueErrors` and perhaps we can prevent all of them with little refactoring?

**Comment by aatle** [comment]
The `else` clause fixes are trivial, with one of three *possibly* using a default instead of a ValueError. I guess they can be rebased, with commits to remove the ignores and add code coverage for them.
However, removing the ignores for some of the other warnings is difficult, they are nasty cases with no obviously better solution.

**Comment by aatle** [comment]
Do you think this PR can be closed now? My main thoughts is that the benefits are small, the code would still give an error either way, just a different type of error, since the changes are just raising explicit errors. The original issue can also probably be closed.

**Comment by DanielNoord** [comment]
Let's close the PR and issue :)

**Pull request closed without merging**